presto-main/src/main/java/com/facebook/presto/execution/scheduler/SqlQueryScheduler.java

```diff
@@ -598,13 +603,13 @@ else if (partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
     private Optional<int[]> getBucketToPartition(
             PartitioningHandle partitioningHandle,
             Function<PartitioningHandle, NodePartitionMap> partitioningCache,
-            Map<PlanNodeId, SplitSource> splitSources,
+            PlanNode fragmentRoot,
             List<RemoteSourceNode> remoteSourceNodes)
     {
         if (partitioningHandle.equals(SOURCE_DISTRIBUTION) || partitioningHandle.equals(SCALED_WRITER_DISTRIBUTION)) {
             return Optional.of(new int[1]);
         }
-        else if (!splitSources.isEmpty()) {
+        else if (PlanNodeSearcher.searchFrom(fragmentRoot).where(node -> node instanceof TableScanNode).findFirst().isPresent()) {
```

Comment:
wow :). Is this the often used pattern in execution? ( looks like built for optimizer~)